# Data consistency checks
95.  Import libraries and data sets
98.  Check and change dtypes
112. Consistency checks df_ords
124. Consistency checks df_prods
146. Export
149. Import and double check departments data set

In [95]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [96]:
# Import data sets
df_ords = pd.read_csv(r'/Users/OldBobJulia/Desktop/CF/Course/4. Python/Instacart Basket Analysis/02 Data/Prepared data/orders_wrangled.csv')

In [97]:
df_prods = pd.read_csv(r'/Users/OldBobJulia/Desktop/CF/Course/4. Python/Instacart Basket Analysis/02 Data/Prepared data/prods_wrangled.csv')

In [98]:
# Check df_ords
df_ords.shape

(3421083, 7)

In [99]:
df_ords.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [100]:
df_ords.dtypes

Unnamed: 0                  int64
order_id                    int64
user_id                     int64
order_number                int64
orders_day_of_week          int64
order_hour_of_day           int64
days_since_prior_order    float64
dtype: object

In [101]:
# Drop Unnamed: 0
df_ords = df_ords.drop(columns = ['Unnamed: 0'])

In [102]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                     NaN  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

In [103]:
# Change back data types as csv does not preserve them 
df_ords['order_id'] = df_ords['order_id'].astype('str')
df_ords['user_id'] = df_ords['user_id'].astype('int16')
df_ords['order_number'] = df_ords['order_number'].astype('str')
df_ords['orders_day_of_week'] = df_ords['orders_day_of_week'].astype('int8')
df_ords['order_hour_of_day'] = df_ords['order_hour_of_day'].astype('int8')
df_ords['days_since_prior_order'] = df_ords['days_since_prior_order'].astype('float16')

In [104]:
df_ords.dtypes

order_id                   object
user_id                     int16
order_number               object
orders_day_of_week           int8
order_hour_of_day            int8
days_since_prior_order    float16
dtype: object

In [105]:
# Check df_prods
df_prods.shape

(49693, 5)

In [106]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [107]:
df_prods.describe()

Unnamed: 0    product_id  department_id        prices
count  49693.0000  49693.000000   49693.000000  49693.000000
mean   24846.0000  24844.345139      11.728433      9.994136
std    14345.2778  14343.717401       5.850282    453.519686
min        0.0000      1.000000       1.000000      1.000000
25%    12423.0000  12423.000000       7.000000      4.100000
50%    24846.0000  24845.000000      13.000000      7.100000
75%    37269.0000  37265.000000      17.000000     11.200000
max    49692.0000  49688.000000      21.000000  99999.000000

In [108]:
# Drop Unnamed: 0
df_prods = df_prods.drop(columns = ['Unnamed: 0'])

In [109]:
df_prods.head()

product_id                                       product_name  \
0           1                         Chocolate Sandwich Cookies   
1           2                                   All-Seasons Salt   
2           3               Robust Golden Unsweetened Oolong Tea   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           5                          Green Chile Anytime Sauce   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [110]:
df_prods.dtypes

product_id         int64
product_name      object
department_id      int64
prices           float64
dtype: object

In [111]:
df_prods.describe()

product_id  department_id        prices
count  49693.000000   49693.000000  49693.000000
mean   24844.345139      11.728433      9.994136
std    14343.717401       5.850282    453.519686
min        1.000000       1.000000      1.000000
25%    12423.000000       7.000000      4.100000
50%    24845.000000      13.000000      7.100000
75%    37265.000000      17.000000     11.200000
max    49688.000000      21.000000  99999.000000

In [112]:
# Df_ords: Check for mixed types

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

There is no mixed type column in df_ords

In [113]:
# Df_ords: Check for missing values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206209 missing values in the column days since prior order. In fact, these should be zeroes because it simply means the customer was a new customer at this point, so they didn't have any prior orders. 

In [114]:
# Impute "0" for missing values in days_since_prior_order
df_ords['days_since_prior_order'].fillna('0', inplace=True)

In [115]:
df_ords.head()

order_id  user_id order_number  orders_day_of_week  order_hour_of_day  \
0  2539329        1            1                   2                  8   
1  2398795        1            2                   3                  7   
2   473747        1            3                   3                 12   
3  2254736        1            4                   4                  7   
4   431534        1            5                   4                 15   

  days_since_prior_order  
0                      0  
1                   15.0  
2                   21.0  
3                   29.0  
4                   28.0

In [116]:
df_ords.isnull().sum()

order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

There are no more missing values in df_ords

In [117]:
# Df_ords: check for duplicates
df_dups = df_ords[df_ords.duplicated()]

In [118]:
df_dups.head()

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

In [119]:
df_dups.shape

(0, 6)

In [120]:
# Double checking no dups
print(df_ords.duplicated().sum())

0


There are no duplicates

In [121]:
# Df_ords: check summary statistics
df_ords.describe()

user_id  orders_day_of_week  order_hour_of_day
count  3.421083e+06        3.421083e+06       3.421083e+06
mean   2.604052e+02        2.776219e+00       1.345202e+01
std    1.849610e+04        2.046829e+00       4.226088e+00
min   -3.276800e+04        0.000000e+00       0.000000e+00
25%   -1.553700e+04        1.000000e+00       1.000000e+01
50%    1.261000e+03        3.000000e+00       1.300000e+01
75%    1.561600e+04        5.000000e+00       1.600000e+01
max    3.276700e+04        6.000000e+00       2.300000e+01

In [122]:
df_ords.shape

(3421083, 6)

In [123]:
# Df_prods: check for mixed types
for col in df_prods.columns.tolist():
    weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_prods[weird]) > 0:
        print (col)

product_name


The product_name column is a mixed type column. It should be string. 

In [124]:
# Set product_name to string
df_prods['product_name'] = df_prods['product_name'].astype('str')

In [125]:
for col in df_prods.columns.tolist():
    weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_prods[weird]) > 0:
        print (col)

In [126]:
# Df_prods: check for missing values 
df_prods.isnull().sum()

product_id       0
product_name     0
department_id    0
prices           0
dtype: int64

There are no missing values

However I remember from doing this exercise earlier that there are in fact NaN in product_name. This could have to do with the change of data type in that column. 

In [127]:
# Try to find those NaN
df_prods[df_prods['product_name'] == 'NaN']

Empty DataFrame
Columns: [product_id, product_name, department_id, prices]
Index: []

In [128]:
df_prods[df_prods['product_name'] == 'Nan']

Empty DataFrame
Columns: [product_id, product_name, department_id, prices]
Index: []

In [129]:
df_prods[df_prods['product_name'] == 'nan']

product_id product_name  department_id  prices
33             34          nan             14    12.2
68             69          nan              7    11.8
115           116          nan              3    10.8
261           262          nan             13    12.1
525           525          nan             11     1.2
1511         1511          nan             16    14.3
1780         1780          nan             11    12.3
2240         2240          nan              1    14.2
2586         2586          nan             13    12.4
3159         3159          nan             11    13.1
3230         3230          nan             16    14.4
3736         3736          nan              8    14.8
4283         4283          nan              7    14.4
4790         4790          nan             16    14.5
38187       38183          nan             12    20.9
40444       40440          nan             16    14.8

In [130]:
# Create subset with missing values only
df_prods_nan = df_prods.loc[df_prods['product_name'] == 'nan']

In [131]:
df_prods_nan.head()

product_id product_name  department_id  prices
33           34          nan             14    12.2
68           69          nan              7    11.8
115         116          nan              3    10.8
261         262          nan             13    12.1
525         525          nan             11     1.2

In [132]:
df_prods_nan.shape

(16, 4)

In [133]:
# Create subset df_prods_clean excluding above subset with nan
df_prods_clean = df_prods.drop(df_prods_nan.index, axis=0)

In [134]:
df_prods_clean.shape

(49677, 4)

I have dropped the 16 rows where product name is "nan"

In [135]:
# Df_prods: check for duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [136]:
df_dups.shape

(5, 4)

There are 5 duplicate rows in df_prods_clean. I will remove them

In [137]:
# Create new subset without duplicate rows
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [138]:
df_prods_clean_no_dups.shape

(49672, 4)

In [139]:
# Df_prods: Check summary statistics
df_prods_clean_no_dups.describe()

product_id  department_id        prices
count  49672.000000   49672.000000  49672.000000
mean   24850.349775      11.728942      9.993282
std    14340.705287       5.850779    453.615536
min        1.000000       1.000000      1.000000
25%    12432.750000       7.000000      4.100000
50%    24850.500000      13.000000      7.100000
75%    37268.250000      17.000000     11.100000
max    49688.000000      21.000000  99999.000000

In [77]:
df_prods_clean_no_dups.shape

(49672, 4)

In [140]:
# Check summary statistics 
df_prods_clean_no_dups.describe()

product_id  department_id        prices
count  49672.000000   49672.000000  49672.000000
mean   24850.349775      11.728942      9.993282
std    14340.705287       5.850779    453.615536
min        1.000000       1.000000      1.000000
25%    12432.750000       7.000000      4.100000
50%    24850.500000      13.000000      7.100000
75%    37268.250000      17.000000     11.100000
max    49688.000000      21.000000  99999.000000

We have the unusually high price of 99999. 

In [141]:
# Check for all outliers price > 100
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] > 100]

product_id                      product_name  department_id   prices
21554       21553  Lowfat 2% Milkfat Cottage Cheese             16  14900.0
33666       33664             2 % Reduced Fat  Milk             16  99999.0

There are 2 products where the price is obviously wrong. This must be corrupted somehow. 

In [142]:
# Assign them 2 Nan for now
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] >100, 'prices'] = np.nan

/opt/anaconda3/lib/python3.8/site-packages/pandas/core/indexing.py:1720: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(loc, value, pi)


In [143]:
# Check again
df_prods_clean_no_dups.loc[df_prods_clean_no_dups['prices'] > 100]

Empty DataFrame
Columns: [product_id, product_name, department_id, prices]
Index: []

In [144]:
df_prods_clean_no_dups.shape

(49672, 4)

In [145]:
df_prods_clean_no_dups.describe()

product_id  department_id        prices
count  49672.000000   49672.000000  49670.000000
mean   24850.349775      11.728942      7.680437
std    14340.705287       5.850779      4.199381
min        1.000000       1.000000      1.000000
25%    12432.750000       7.000000      4.100000
50%    24850.500000      13.000000      7.100000
75%    37268.250000      17.000000     11.100000
max    49688.000000      21.000000     25.000000

In [146]:
# Export
path = r'/Users/OldBobJulia/Desktop/CF/Course/4. Python/Instacart Basket Analysis/02 Data'

In [147]:
df_ords.to_csv(os.path.join(path, 'Prepared Data', 'orders_checked.csv'))

In [148]:
df_prods_clean_no_dups.to_csv(os.path.join(path, 'Prepared Data', 'prods_checked.csv'))

In [149]:
# Import departments data set for consistency check 
df_departments = pd.read_csv(r'/Users/OldBobJulia/Desktop/CF/Course/4. Python/Instacart Basket Analysis/02 Data/Prepared data/departments_wrangled.csv')

In [150]:
df_departments.shape

(21, 2)

In [151]:
df_departments.head()

Unnamed: 0 department
0           1     frozen
1           2      other
2           3     bakery
3           4    produce
4           5    alcohol

In [152]:
df_departments.describe()

Unnamed: 0
count   21.000000
mean    11.000000
std      6.204837
min      1.000000
25%      6.000000
50%     11.000000
75%     16.000000
max     21.000000

In [154]:
# Drop column Unnamed: 0 
df_departments = df_departments.drop(columns = ['Unnamed: 0'])

In [155]:
df_departments.head()

department
0     frozen
1      other
2     bakery
3    produce
4    alcohol

In [156]:
# Check for duplicates 
df_dups_3 = df_departments[df_departments.duplicated()]

In [157]:
df_dups_3.shape

(0, 1)

In [158]:
# Check for missing data 
df_departments.isnull().sum()

department    0
dtype: int64

In [161]:
# Check counts 
df_departments['department'].value_counts(dropna=False)

babies             1
snacks             1
produce            1
beverages          1
breakfast          1
bakery             1
bulk               1
other              1
household          1
deli               1
dairy eggs         1
canned goods       1
frozen             1
pantry             1
meat seafood       1
international      1
missing            1
pets               1
alcohol            1
personal care      1
dry goods pasta    1
Name: department, dtype: int64

Df_departments seems fine as it is. I will leave it as it is. 